### DISCUSSION TITLE
Update Prime Tween Inside Custom Tick (Manual update, UpdateType.Manual)

### DISCUSSION THREAD

**Discussion by sergiyha** [KyryloKuzyk/PrimeTween#135] [General] [answered]
Hi.
I am developing the game utilizing custom ECS engine and have my own tick. I do not have any unity Update() event functions in my project and I want to be confident about the execution order. The question is next: can I invoke the global Prime update inside my tick, it would be a separate system for doing it. Thanks.

**Comment by PalliativeX** [comment]
It would also be great to be able to update manually not the whole tween manager, but separately stored tweens, because for UI it's ok that tweens are updated automatically, but when using tweens for the camera and other game logic it's important to control the execution order.

**Comment by KyryloKuzyk** [comment]
Hi!

I'm afraid PrimeTween wasn't designed with this use case in mind and there is currently no way to update PrimeTween manually. I'm not sure if adding this functionality will benefit a broader audience, but PrimeTween comes with full source code so adding this feature yourself should be relatively easy. 

But maybe knowing the details of how PrimeTween updates its animation can work as an alternative solution for you so you can be confident when things happen:
- PrimeTween updates all its animations in Update() with the script execution order of `-2000`. So any script with a higher execution order than -2000 can safely assume that all PrimeTween animations are already updated.
- Also, animations with specified `startValue` receive their first update in the LateUpdate() after they are created. This guarantees that all animations are displayed at their `startValue` when the frame is rendered. So, if an animation is created before PrimeTween's LateUpdate(), your ECS system can be sure that it will receive its first update before the frame is rendered.

**Reply by TylerTemp** [reply]
Hi, I'm also interested in this topic. I don't care whether I need to update the Manager or individual Tween, but can you point a direction that how can I manually tick it? Like, if the tweenManager has a `Update` function that read `Time.time`? If so I may find a way to init the manager then use reflection or so to change time and update it

**Reply by KyryloKuzyk** [reply]
@TylerTemp Because PrimeTween comes with the full source code, I would say the easiest way to support custom tick is to modify the PrimeTweenManager class.

To modify the source code, first embed PrimeTween into your project: https://discussions.unity.com/t/how-to-locally-modify-source-code-in-a-package/919965

There are 3 methods in which PrimeTween updates all its animations: Update(), LateUpdate(), and FixedUpdate(). Just rename those methods (so Unity doesn't call them in its event loop) and make them public. Then call these methods at appropriate times for your game.

You might also want to add custom support for `float deltaTime, float unscaledDeltaTime` parameters if your game doesn't use Unity's time.

**Comment by KyryloKuzyk** [comment, marked as answer]
Hey @sergiyha @TylerTemp, good news!

I added an experimental feature that lets you update tweens manually. To enable it, add the `PRIME_TWEEN_EXPERIMENTAL` to the ‘Project Settings / Player / Scripting Define Symbols’.

Then you'll be able to manually update PrimeTweenManager from a script:
```csharp
void Update() => PrimeTweenConfig.ManualUpdate(UpdateType.Update);

void LateUpdate() {
    PrimeTweenConfig.ManualUpdate(UpdateType.LateUpdate);
    PrimeTweenConfig.ManualUpdateApplyStartValues(UpdateType.Update);
    PrimeTweenConfig.ManualUpdateApplyStartValues(UpdateType.LateUpdate);
}

void FixedUpdate() => PrimeTweenConfig.ManualUpdate(UpdateType.FixedUpdate);
```
You can also provide custom `deltaTime` and `unscaledDeltaTime` to the `PrimeTweenConfig.ManualUpdate()` method if you have your project doesn't rely on Unity's time.

@PalliativeX Good news for you too 😀
I added a new `UpdateType` parameter to `TweenSettings`, so you can choose to update tweens in LateUpdate. Will that help you with your execution order request?

**Reply by ghost** [reply]
Hi, can the update settings be applied per Tween instead of a global config? Use case is for multiplayer networked games. In netcode, usually the simulation is done in a separate update mode, the network ticks, which can be configured to any arbitrary tick rate like 30 or 60. The UI, of course, runs in Update as it doesn't need to be networked.

Rollback happens in predicted netcode too, so it also important to be able to revert to a previous time. I currently use LitMotion which supports [directly setting the progress](https://github.com/annulusgames/LitMotion/issues/132#issuecomment-2561565024), manually step the tween, whether it be forwards or backwards.

**Reply by KyryloKuzyk** [reply]
@nscimerical PrimeTween has `Tween.progress` and `Tween.elapsedTime` properties, which you can modify manually. Do you think that would help?

**Reply by ghost** [reply]
Oh yes, that actually does. I apologize, I completely missed that one

**Reply by jollypunch** [reply]
Would it be possible to enable manual updating by tween/sequence? E.g. Having most tweens function normally, or even be updated with ManualUpdate, and having some tweens that are only updated via elapsedTime or elapsedTotalTime? Otherwise the manager keeps updating those and messes things up.

At the moment it seems only one or the other is possible for all tweens, e.g. by disabling the manager altogether with one ManualUpdate, but that is a big hindrance for tweening UI etc., like Palliative mentioned.

A good solution might be to add UpdateType.Manual.

**Reply by KyryloKuzyk** [reply]
Hey @jollypunch, I'm not entirely sure I understand the use case of UpdateType.Manual. Can you please elaborate on why you would want to update elapsedTime manually?

One solution I see for you is to set `isPaused = true`, then PrimeTweenManager will not update such tweens. Then you can manually update elapsedTime independently of other animations. One caveat to have in mind is that paused animations are not stopped automatically when they reach the end, so you'll have to Stop them manually too.

**Reply by jollypunch** [reply]
A classic use case is for online games that need deterministic animations where colliders can change positions/size. This means updating via tick and being able to rewind/reproduce at will. I embedded the package and I've been working with UpdateType.Manual, it's working well.
